File: lr/src/questions/q_0128_longest_consecutive_sequence.rs
use std::collections::HashSet;

pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
    if nums.is_empty() {
        return 0;
    }

    let mut result = 0;
    let nums_set: HashSet<&i32> = HashSet::from_iter(nums.iter());

    for n in &nums {
        if nums_set.get(&(n - 1)).is_none() {
            let mut temp_count = 1;
            let mut temp = *n;
            while nums_set.get(&(temp + 1)).is_some() {
                temp_count += 1;
                temp += 1;
            }
            if temp_count > result {
                result = temp_count;
            }
        }
    }

    result
}

#[test]
fn test_longest_consecutive() {
    assert_eq!(longest_consecutive(vec![]), 0);
    assert_eq!(longest_consecutive(vec![100, 4, 200, 1, 3, 2]), 4);
    assert_eq!(longest_consecutive(vec![0, 3, 7, 2, 5, 8, 4, 6, 0, 1]), 9);
}

File: lg/src/questions/q_0128_longest_consecutive_sequence.go
package questions

func longestConsecutive(nums []int) int {
	numsMap := make(map[int]struct{}, len(nums))
	for _, num := range nums {
		numsMap[num] = struct{}{}
	}
	result := 0
	for _, num := range nums {
		if _, ok := numsMap[num-1]; !ok {
			tempCount := 0
			nextNum := num
			for {
				if _, ok := numsMap[nextNum]; ok {
					tempCount += 1
					nextNum += 1
				} else {
					break
				}
			}
			if tempCount > result {
				result = tempCount
			}
		}
	}
	return result
}
